compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -692,32 +674,47 @@ impl<'tcx> Cx<'tcx> {
                     };
 
                     let source = if let Some((did, offset, var_ty)) = var {
-                        let mk_const = |literal| Expr {
-                            temp_lifetime,
-                            ty: var_ty,
-                            span: expr.span,
-                            kind: ExprKind::Literal { literal, user_ty: None, const_id: None },
-                        };
-                        let offset = self.thir.exprs.push(mk_const(ty::Const::from_bits(
+                        let param_env_ty = self.param_env.and(var_ty);
+                        let size = self
+                            .tcx
+                            .layout_of(param_env_ty)
+                            .unwrap_or_else(|e| {
+                                panic!("could not compute layout for {:?}: {:?}", param_env_ty, e)
+                            })
+                            .size;
+                        let const_literal = ty::Const::from_bits(
                             self.tcx,
                             offset as u128,
                             self.param_env.and(var_ty),
-                        )));
+                        );
+                        match const_literal.val() {
+                            ty::ConstKind::Value(ConstValue::Scalar(Scalar::Int(int))) => {
+                                assert_eq!(int.size(), size);
```

Comment:
I still think that the way we create the `ScalarInt`s must be the source for this problem. Note that this assertion is not triggered when compiling core.